consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,488 @@
+consensus. The application also provides application specific implementations
+for certain functions that the consensus algorithm relies upon, such as a
+function that determines the validity of a value. 
+
+### Algorithm
+
+See supporting [functions](#Functions) and [notation](#Appendix-1-Notation).
+
+Note all state modifications are contained in the following code block,
+supporting functions cannot modify state, they all operate without side
+effects.
+
+Variable names                  |Instance state                          
+--------------------------------|----------------------------------------
+`H - height`                    |`Hc - current height`                                      
+`R - round          `           |`Rc - current round`                    
+`V - value`                     |`Rd - desired round`                    
+`T - Message type`              |`Vc - currently proposed value`         
+`RCC - round change certificate`|`PCc - current prepared certificate`    
+`PC - prepared certificate`     |`Sc - current participant state`        
+`S - participant state`         |
+`M or N - message sets`         |
+
+```
+
+// Upon receiving this event the algorithm transitions to the next height the
+// final committed event signifies that the application has accepted the agreed
+// upon value for the current height.
+upon: <FinalCommittedEvent>
+  Hc ← Hc+1
+  Rc ← 0
+  Rd ← 0
+  Sc ← AcceptRequest
+  Vc ← nil
+  schedule onRoundChangeTimeout(Hc, 0) after roundChangeTimeout(0)
+
+// A request event is a request to reach agreement on the provided value, the
+// request event is sent by the application, if this parcicipant is the proposer
+// it will propose that value by sending a preprepared message.
+upon: <RequestEvent, Hc, V> && Sc = AcceptRequest
+  if Rc = 0 && isProposer(Hc, Rd) {
+    bc(<Preprepare, Hc, 0, V, nil>)
+  }
+
+// When receiving a preprepare from a proposer participants will vote for the
+// value (if valid) by sending a prepare message.
+upon: <Preprepare, Hc, Rd, V, RCC> from proposer(Hc, Rd) && Sc = AcceptRequest
+  if (Rd > 0 && validRCC(V, RCC)) || (Rd = 0 && RCC = nil)  {
+    Rc ← Rd
+    Vc ← V
+    Sc ← Preprepared
+    bc(<Prepare, Hc, Rd, Vc>)
+  }
+
+// When a participant sees at least 2f+1 prepare or commit messages for a value
+// it will send a commit message for that value.
+upon: M ← { <T, Hc, Rd, Vc> : T ∈ {Prepare, Commit} } && |M| >= 2f+1 && Sc ∈ {AcceptRequest, Preprepared} 
+  Sc ← Prepared
+  PCc ← <PreparedCertificate, M, Vc>
+  bc(<Commit, Hc, Rd, Vc>)
+
+// When a participant sees at least 2f+1 commit messages for a value, they
+// consider that value committed (agreed) and pass the value to the application,
+// which will in turn issue a final committed event if the value is considered
+// valid by the application.
+upon: M ← { <Commit, Hc, Rd, Vc> } && |M| >= 2f+1 && Sc ∈ {AcceptRequest, Preprepared, Prepared} 
```

Comment:
Is the same message set? for commits and prepares? I think we need a better explanation of what the message set are. 

2. Also, why can Sc be in all these states?
3. Why `Rd` and not `Rc`, assuming they can be different, right? I would expect this to be `Rc` since that's the current round we are trying to accumulate messages for

Reply:
>Is the same message set? for commits and prepares? I think we need a better explanation of what the message set are.

I'm going to add a variable to clarify the message set.

>2. Also, why can Sc be in all these states?

Its a mistake, it should only be Preprepared and Prepared

>3. Why Rd and not Rc, assuming they can be different, right? I would expect this to be Rc since that's the current round we are trying to accumulate messages for

I don't know, thats what is in the code: https://github.com/celo-org/celo-blockchain/blob/be8e779fb879e63cfae8f0d60f9b4f9ff3163fd3/consensus/istanbul/core/backlog.go#L63-L71

I don't actually think we need to have desired round and current round, we could just do with one variable, you can see this relatively easily by looking at the spec and seeing where they would differ. We can go over this together if you're interested. I think removing desired round would make the protocol much easier to understand.

Reply:
I remember desired round be a corner case. When the val set are in diff rounds. Desired round helped to speed up the consensus on which is the current round

Reply:
I've scheduled some time tomorrow to go over this, seems like its quite nuanced